### ISSUE TITLE
CopyToOutputDirectory Does Not Respect PreserveNewest and TargetPath

### ISSUE THREAD

**Issue by BenVillalobos** [dotnet/msbuild#6496]
Hello @BenVillalobos.  Now that version 16.10 has been released, I am able to test.  Unfortunately, I think this change exchanged one problem for another.

```xml
<Content Include="Files\**">
  <Link>Files\%(Filename)%(Extension)</Link>
  <TargetPath>%(Filename)%(Extension)</TargetPath>
  <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
</Content>
```

I have verified the correctness of the first two lines (i.e. the  and  elements).  However, the last line with the  element is not working.  After completing the first build, if I tell Visual Studio to build again, then it builds the project instead of correctly saying that it is update to date.

I tested by modifying [the MWE](https://www.dropbox.com/s/h9vk0mhgd3076tz/ClassLibrary1.zip?dl=0) I provided in [my original VSDC isssue](https://developercommunity.visualstudio.com/t/copytooutputdirectorypreservenewest-ignored-inside/1332219).

_Originally posted by @TysonMN in https://github.com/dotnet/msbuild/issues/6237#issuecomment-849618338_

**Comment by BenVillalobos** [comment]
I wonder if CopyToOutputDirectory checks an item's `Link` metadata for up-to-date-ness. Need to get command line binlogs.

**Comment by rainersigwald** [comment]
This could also be a problem with the project system's [fast up-to-date check](https://github.com/dotnet/project-system/blob/main/docs/up-to-date-check.md). Might be interesting to up their verbosity (tools -> options -> Projects and Solutions -> SDK-style Projects -> Up to Date Checks -> Logging Level) and see what they say.

**Comment by TysonMN** [comment]
@BenVillalobos, thanks for creating this issue.  As a nitpick, the copy paste didn't work correctly.  There is some information missing.  I avoid that problem by quoting the comment (i.e. click the three dots in the top-right corner and then select "Quote reply"), deleting the initial right angle brackets and spaces, and then copying that text.

**Comment by BenVillalobos** [comment]
@TysonMN Weird, that seems to be an issue with `3 dots` -> `Reference in new issue`

**Comment by TysonMN** [comment]
Great suggestion @rainersigwald.  Here are the relevant portions of those logs (after optimizing the project structure for smaller logs).
```
1>FastUpToDate: Checking PreserveNewest file 'C:\A\Files\TextFile.txt': (ClassLibrary1)
1>FastUpToDate:     Source 5/27/2021 7:54:59 AM: 'C:\A\Files\TextFile.txt'. (ClassLibrary1)
1>FastUpToDate: Destination 'C:\A\bin\Debug\netstandard2.0\Files\TextFile.txt' does not exist, not up to date. (ClassLibrary1)
```

The fast up-to-date check did not take into account the information given in the `TargetPath` element.

Specifically, it checked
```
C:\A\bin\Debug\netstandard2.0\Files\TextFile.txt
```
when it should have checked
```
C:\A\bin\Debug\netstandard2.0\TextFile.txt
```

**Comment by BenVillalobos** [comment]
@TysonMN Aha! Does this also repro on a command line build?

**Comment by TysonMN** [comment]
I don't think so.  It seems to me like the project is being built.  Here is the output.  Do you have the same conclusion?  If this is output is inconclusive, is there some why I can make it more conclusive (i.e. enable some more logging)?

```
C:\A>dir
 Volume in drive C has no label.
 Volume Serial Number is EEAA-DB41

 Directory of C:\A

05/27/2021  02:10 PM    <DIR>          .
05/27/2021  02:10 PM    <DIR>          ..
02/05/2021  04:17 PM                80 Class1.cs
05/27/2021  01:44 PM               395 ClassLibrary1.csproj
05/27/2021  01:55 PM             1,131 ClassLibrary1.sln
05/27/2021  01:55 PM    <DIR>          Files
               3 File(s)          1,606 bytes
               3 Dir(s)  23,863,943,168 bytes free

C:\A>dotnet build
Microsoft (R) Build Engine version 16.10.0+4242f381a for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored C:\A\ClassLibrary1.csproj (in 307 ms).
  ClassLibrary1 -> C:\A\bin\Debug\netstandard2.0\ClassLibrary1.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:03.49

C:\A>dotnet build
Microsoft (R) Build Engine version 16.10.0+4242f381a for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  ClassLibrary1 -> C:\A\bin\Debug\netstandard2.0\ClassLibrary1.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.36
```

**Comment by BenVillalobos** [comment]
@TysonMN The best way to provide logs for us is with binary logs: aka.ms/msbuild/binlog

It'll output a `.binlog` file that you can share. Be aware it will include things like environment variables.

~~Weird that this class library project always builds, even if I remove everything but the target framework. Meanwhile creating a new `dotnet new classlib` and building twice doesn't build the second time (all on command line).~~

~~I do see `Input file "obj\Debug\netstandard2.0\ClassLibrary1.GeneratedMSBuildEditorConfig.editorconfig" is newer than output file "obj\Debug\netstandard2.0\ClassLibrary1.dll".` is the reason why the project is running CoreCompile (and therefore building). Which is running because the editorconfig file is always generated.~~ But when debugging with @rainersigwald it magically stopped ._.

**Comment by BenVillalobos** [comment]
Transferring to project system. We think the fast up to date check has to be updated to account for how MSBuild respects `TargetPath` metadata. We modified `AssignTargetPaths` to respect `TargetPath` when it's already set, where previously the `Link` metadata would take priority. See https://github.com/dotnet/msbuild/pull/6237
